crypto/evp/m_sigver.c

```diff
@@ -43,59 +44,74 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
         ctx->provctx = NULL;
     }
 
-    if (ctx->pctx == NULL) {
+    if (ctx->pctx == NULL)
         ctx->pctx = EVP_PKEY_CTX_new(pkey, e);
-        if (ctx->pctx == NULL)
-            return 0;
-    } else if (pkey != NULL) {
-        if (!EVP_PKEY_up_ref(pkey))
-            return 0;
-        EVP_PKEY_free(ctx->pctx->pkey);
-        ctx->pctx->pkey = pkey;
-    }
```

Comment:
I'm confused by the removal of this `else if` block. Why is this no longer needed?

Reply:
It seriously changes the semantics.  Here's what the 1.1.1 code does:

https://github.com/openssl/openssl/blob/689c07b8d502301d82e09c1db04daf3c13b1d35d/crypto/evp/m_sigver.c#L28-L45

So `pkey` is *only* assigned to the PKEY context if it's created here, otherwise, it's *only* used to set `type` if that one is NULL, and only under specific conditions.

Spuriously assigning an *already existing* PKEY context requires more work (there's otherwise a great risk of a mismatch between the new key and the rest of the PKEY context data), to the point that it's better to simply free the context and re-create it, and I think that would be a sore surprise for the caller.

Reply:
Mind you, if you look below, I'm using `type` or `pkey` to set `mdname` if the latter is NULL without looking at more conditions...  that broadens the conditions to set the MD type internally, but I think that's within safe limits.

Reply:
> It seriously changes the semantics. 

Looking at the history of this I have no idea why I added that. So, ok.